presto-main/src/main/java/com/facebook/presto/cost/PlanNodeCostEstimate.java

```diff
@@ -13,46 +13,52 @@
  */
 package com.facebook.presto.cost;
 
+import com.google.common.primitives.Doubles;
+
 import java.util.Objects;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
 import static java.lang.Double.NaN;
 import static java.lang.Double.POSITIVE_INFINITY;
 import static java.lang.Double.isNaN;
 
 public final class PlanNodeCostEstimate
 {
-    public static final PlanNodeCostEstimate INFINITE_COST = new PlanNodeCostEstimate(POSITIVE_INFINITY, POSITIVE_INFINITY, POSITIVE_INFINITY);
-    public static final PlanNodeCostEstimate UNKNOWN_COST = new PlanNodeCostEstimate(NaN, NaN, NaN);
-    public static final PlanNodeCostEstimate ZERO_COST = new PlanNodeCostEstimate(0, 0, 0);
+    public static final PlanNodeCostEstimate INFINITE_COST = new PlanNodeCostEstimate(POSITIVE_INFINITY, POSITIVE_INFINITY, POSITIVE_INFINITY, POSITIVE_INFINITY);
+    public static final PlanNodeCostEstimate UNKNOWN_COST = new PlanNodeCostEstimate(NaN, NaN, NaN, NaN);
+    public static final PlanNodeCostEstimate ZERO_COST = new PlanNodeCostEstimate(0, 0, 0, 0);
 
     public static PlanNodeCostEstimate cpuCost(double cpuCost)
     {
-        return new PlanNodeCostEstimate(cpuCost, 0, 0);
+        return new PlanNodeCostEstimate(cpuCost, 0, 0, 0);
     }
 
     public static PlanNodeCostEstimate memoryCost(double memoryCost)
     {
-        return new PlanNodeCostEstimate(0, memoryCost, 0);
+        return new PlanNodeCostEstimate(0, memoryCost, memoryCost, 0);
     }
 
     public static PlanNodeCostEstimate networkCost(double networkCost)
     {
-        return new PlanNodeCostEstimate(0, 0, networkCost);
+        return new PlanNodeCostEstimate(0, 0, 0, networkCost);
     }
 
     private final double cpuCost;
-    private final double memoryCost;
+    private final double peakMemoryCost;
+    private final double peakMemoryWhenOutputting;
```

Comment:
Sorry for not giving this explanation upfront.

So, the basic idea is as outlined in https://github.com/prestodb/presto/pull/11495#issuecomment-422558916: to calculate peak memory usage of a query.
To calculate this recursively, we need to know for every node in a plan tree:
- peak memory usage of the subtree
- peak memory usage when the node is outputting (producing output) -- this is required to compute peak memory usage of plan tree enclosing given node
  - this is based on observation that until given node is outputting, plan nodes higher up need no memory (they have nothing to process)
  - also, some nodes will consume all input before outputting, so when they are outputting, their subtrees no longer need memory
- the two properties of source nodes -- _peak memory usage_ and _peak memory usage when the node is outputting_ -- combined with node-specific formulas should allow to computing the properties higher up, realizing _peak memory usage_ calculation for the whole plan

Now, the ugly part. In current cost calculation code, PlanNodeCostEstimate conveniently represents two things:
- local cost of the node
- total cost of the plan subtree

It's no wonder we reuse single class for this, as currently total cost of the plan subtree is just a sum of local costs of all nodes.
After this change, this design no longer fully fits:
- when calculating local cost of the node, we should calculate `[cpu, memory, network]`
- when calculating total cost of the plan subtree, we should calculate `[cpu, peak memory, network]` (this can still be the same type as local cost of the node), but we temporarily need to hold `[cpu, peak memory, peak memory when outputting, network]`
  - it might be that the "extended" type is just an inner thing is `CostProvider` implementation, so maybe the overall design still holds
- summing cost vectors does not always make sense

For now, I didn't address the design part at all. I am just conveniently reusing `PlanNodeCostEstimate` everywhere.